src/hotspot/share/opto/superword.cpp

```diff
@@ -1326,6 +1364,44 @@ bool SuperWord::independent(Node* s1, Node* s2) {
   return independent_path(shallow, deep);
 }
 
+//------------------------------find_dependence---------------------
+// Is any s1 in p dependent on any s2 in p? Yes: return such a s2. No: return nullptr.
+// We could query independent(s1, s2) for all pairs, but that results
+// in O(p.size * p.size) graph traversals. We can do it all in one BFS!
+// Start the BFS traversal at all nodes from the pack. Traverse DepPreds
+// recursively, for nodes that have at least depth min_d, which is the
+// smallest depth of all nodes from the pack. Once we have traversed all
+// those nodes, and have not found another node from the pack, we know
+// that all nodes in the pack are independent.
+Node* SuperWord::find_dependence(Node_List* p) {
+  if (p->at(0)->is_reduction()) {
+    return nullptr; // ignore reductions
+  }
+  ResourceMark rm;
+  Unique_Node_List worklist; // traversal queue
+  int min_d = depth(p->at(0));
+  visited_clear();
+  for (uint k = 0; k < p->size(); k++) {
+    Node* n = p->at(k);
+    min_d = MIN2(min_d, depth(n));
+    worklist.push(n); // start traversal at all nodes in p
+    visited_set(n); // mark node
+  }
+  for (uint i = 0; i < worklist.size(); i++) {
+    Node* n = worklist.at(i);
+    for (DepPreds preds(n, _dg); !preds.done(); preds.next()) {
+      Node* pred = preds.current();
+      if (in_bb(pred) && depth(pred) >= min_d) {
```

Comment:
Hi @eme64 , Dependency captures three scenarios :-
  a) RAW(ture) :  Store -> Load
  b) WAR(anti) :  Load -> Store
  c)  WAW        :  Store -> Store.
RAR is a non-dependency.    Existing routine independent_path  is called during statement packing after its certain that statements are isomorphic.  Newly added find_dependence routine optimizes the dependency checking but it does not seem to be skipping over anti-dependence checks which may prevent vectorization in some cases.  Can you share your thoughts on this.

Reply:
Thanks @jatin-bhateja . In essence, I am trusting `DepPreds` to give me the correct dependency information. Is that not correct? Do you have an example where `DepPreds` would have a `Load` depend on a `Load`, without a store in between?

Reply:
You are correct Load-Load(RAR) is not captured in dependence graph anyways.

My above comment was in context of scenario b),  I verified that newly added find_dependence method will not find any circular intra-pack dependence in this case.

Reply:
@jatin-bhateja Ah, I misunderstood. Thanks for clarifying. So you verified that WAR is not an issue? Thanks!

I agree that it should not be an issue. For a `Load -> Store` to block vectorization, we would have to find a dependency path from a `Load -> Store -> Load` (LSL) or a `Store -> Load -> Store` (SLS), since my verification only checks if any of the memops in the pack depend on another memop of the pack (both loads or both stores).
But both LSL and SLS have a `Store -> Load` (RAW) and so we cannot vectorize.
Do you agree with this argument?

Reply:
> @jatin-bhateja Ah, I misunderstood. Thanks for clarifying. So you verified that WAR is not an issue? Thanks!
> 
> I agree that it should not be an issue. For a `Load -> Store` to block vectorization, we would have to find a dependency path from a `Load -> Store -> Load` (LSL) or a `Store -> Load -> Store` (SLS), since my verification only checks if any of the memops in the pack depend on another memop of the pack (both loads or both stores). But both LSL and SLS have a `Store -> Load` (RAW) and so we cannot vectorize. Do you agree with this argument?

Correct, as long as Load in (Store->Load) not feeding back into any other Store in the same pack. 

Without complicating the way I see this is that dependence checking looking for an intra-pack dependency where pack is comprised of isomorphic nodes (loads/stores) does a walk over data pendency edges each memory operation in the pack and if the traversal lands back to one the nodes in the pack it should declare a cyclic dependence.